### PR TITLE
MINIFICPP-2121 - Use std::atomic_flag instead of semaphore

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: minifi_main/MiNiFiMain.cpp [hunk 1]
```diff
@@ -91,27 +90,26 @@ static sem_t *process_running;
...
+    process_running.wait(false);
     return TRUE;
   }
   return FALSE;
 }
 
 void SignalExitProcess() {
-  sem_post(flow_controller_running);
+  flow_controller_running.clear();
+  flow_controller_running.notify_all();
```
Comment (szaszm) [inline minifi_main/MiNiFiMain.cpp:105]: Is this approach signal safe?

Comment (adamdebreceni) [reply minifi_main/MiNiFiMain.cpp:105]: according to the standard calling a non-static member function of `std::atomic_flag` is a "plain lock-free atomic operation" which is allowed in signal handlers

File: minifi_main/MiNiFiMain.cpp [hunk 2]
```diff
@@ -91,27 +90,26 @@ static sem_t *process_running;
...
   if (signal == CTRL_C_EVENT || signal == CTRL_BREAK_EVENT) {
-    int ret = ETIMEDOUT;
-    while (ret == ETIMEDOUT) {
-      if (flow_controller_running) { sem_post(flow_controller_running); }
-      const struct timespec timeout_100ms { .tv_sec = 0, .tv_nsec = 100000000};
-      ret = sem_timedwait(process_running, &timeout_100ms);
-    }
+    flow_controller_running.clear();
+    flow_controller_running.notify_all();
+    process_running.wait(false);
```
Comment (fgerlits) [inline minifi_main/MiNiFiMain.cpp:97]: should this be `wait(true)`?  from the docs, it looks to me that `wait(x)` waits for the flag to be _not_ x, and we want to wait for `process_running` to become `false` (ie, not `true`)

Comment (adamdebreceni) [reply minifi_main/MiNiFiMain.cpp:97]: good catch, indeed that seems to be the case

**Review by lordgamez — APPROVED** [review]

**Review by fgerlits — APPROVED** [review]

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**